osu.Game/Database/StableImportManager.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using osu.Framework;
+using osu.Framework.Allocation;
+using osu.Framework.Extensions.EnumExtensions;
+using osu.Framework.Graphics;
+using osu.Framework.Platform;
+using osu.Game.Beatmaps;
+using osu.Game.Collections;
+using osu.Game.IO;
+using osu.Game.Overlays;
+using osu.Game.Overlays.Settings.Sections.Maintenance;
+using osu.Game.Scoring;
+using osu.Game.Skinning;
+
+namespace osu.Game.Database
+{
+    public class StableImportManager : Component
+    {
+        [Resolved]
+        private SkinManager skins { get; set; }
+
+        [Resolved]
+        private BeatmapManager beatmaps { get; set; }
+
+        [Resolved]
+        private ScoreManager scores { get; set; }
+
+        [Resolved]
+        private CollectionManager collections { get; set; }
+
+        [Resolved]
+        private OsuGame game { get; set; }
+
+        [Resolved]
+        private DialogOverlay dialogOverlay { get; set; }
+
+        [Resolved(CanBeNull = true)]
+        private DesktopGameHost desktopGameHost { get; set; }
+
+        private StableStorage cachedStorage;
+
+        public bool SupportsImportFromStable => RuntimeInfo.IsDesktop;
+
+        public async Task ImportFromStableAsync(StableContent content)
+        {
+            var stableStorage = await getStableStorage().ConfigureAwait(false);
+            var importTasks = new List<Task>();
+
+            Task beatmapImportTask = Task.CompletedTask;
+            if (content.HasFlagFast(StableContent.Beatmaps))
+                importTasks.Add(beatmapImportTask = beatmaps.ImportFromStableAsync(stableStorage));
+
+            if (content.HasFlagFast(StableContent.Skins))
+                importTasks.Add(skins.ImportFromStableAsync(stableStorage));
+
+            if (content.HasFlagFast(StableContent.Collections))
+                importTasks.Add(beatmapImportTask.ContinueWith(_ => collections.ImportFromStableAsync(stableStorage), TaskContinuationOptions.OnlyOnRanToCompletion));
+
+            if (content.HasFlagFast(StableContent.Scores))
+                importTasks.Add(beatmapImportTask.ContinueWith(_ => scores.ImportFromStableAsync(stableStorage), TaskContinuationOptions.OnlyOnRanToCompletion));
+
+            await Task.WhenAll(importTasks.ToArray()).ConfigureAwait(false);
+        }
+
+        private async Task<StableStorage> getStableStorage()
+        {
+            var stableStorage = game.GetStorageForStableInstall();
+            if (stableStorage != null)
+                return stableStorage;
+
+            if (cachedStorage != null)
+                return cachedStorage;
+
+            var taskCompletionSource = new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
I've never actually seen this `RunContinuationsAsynchronously` flag being used anywhere in the wild. Do you mind explaining why it's here? Is it something to do with `ConfigureAwait()`?

Reply:
If my memory serves me right, when awaiting upon a `TaskCompletionSource` in an async method, the rest of the method will run synchronously on the thread that called any of the `Set*` methods on the TCS; But really,  I added this just to get an additional safeguard that the rest of the method will run asynchronously,  since `SetResult` gets called on the Update thread; but I'm not so confident about this change so I'm fine with removing the flag if it is deemed unecessary.

Reply:
I see. To be quite fair I'm really having trouble even figuring out what code will run on which thread with this sort of complex TPL usage without actually going in and running it...

When eyeballing it, it does seem correct, given that blocks between `await`s are internally converted into continuations, and the next `await` after `getStableStorage()` is that `Task.WhenAll()` that's going to be pretty huge. Probably best that I go and check at some point though, just need to sort out a decent stable directory to test importing with.